### PR TITLE
Enable soft-deletes by default on 7.0.0 or later

### PR DESCRIPTION
**Pull request by dnhatn**
This change enables soft-deletes by default on ES 7.0.0 or later.

### CODE REVIEW

**Review by bleskes — APPROVED** [review]
Lgtm but there are two things that I wanted to double check:
1) is there no documentation change needed?
2) did you look at the consequences of enabling soft deletes regardless of the index version? I can see how  it will make some ops based recoveries fail and go back to file copy but I think that will be rare enough to consider an unconditional default that holds for old data as well.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Now that this is the default there are several places in the CCR docs where this can be cleaned up. Also, the CCR requirements docs should be re-worded.

**Comment by bleskes** [comment]
> An engine with soft-deletes disabled can't open a soft-deletes Lucene index.
I take it you mean the 6.x node will not see that soft deletes are enabled because it has a different default?

>  CCR requires soft-deletes enabled on the leader before following. 
This is a problem in your solution too?

> We need to set soft-deletes default setting explicitly when creating a new index in MetaDataCreateIndexService#IndexCreationTask

I think it's good to explore this (although, I don't think it's enough). My goal is to discuss what the implications would be to make 7.x engines always run with soft deletes, so we can potentially remove the translog retention logic (beyond the global checkpoint aspects). I'm still debating this personally but I wanted to raise the discussion.

**Comment by s1monw** [comment]
> An engine with soft-deletes disabled can't open a soft-deletes Lucene index. If a new index is created in a mixed cluster (7.0 and 6.x) and a 6.x replica won't be able to open a 7.0 soft-deletes index.

can you explain why this is the case?

**Comment by dnhatn** [comment]
@s1monw If we use an implicit default value, it's possible to have a primary on 7.0 with soft-deletes enabled (by default), and a replica on 6.x with soft-deletes disabled (also by default). Then that replica won't be able to open a Lucene index copied from the primary because we set soft-deletes field in IndexWriterConfig [iff soft-deletes is enabled](https://github.com/elastic/elasticsearch/blob/05d52b222f354f53d705a71beadc70f1bad9d1e3/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L2111-L2112). Setting soft-deletes setting explicitly when creating an index would solve these two issues entirely.

**Comment by dnhatn** [comment]
@bleskes I have updated this PR to set the soft-deletes setting explicitly when creating a new index.
@jasontedor I have adjusted the ccr doc.

Could you please have another look? Thank you!

**Comment by s1monw** [comment]
> IndexWriterConfig iff soft-deletes is enabled. Setting soft-deletes setting explicitly when creating an index would solve these two issues entirely.

yeah so I think we don't have to do that. We can just always set it and then we don't have any issues?

**Comment by dnhatn** [comment]
> yeah so I think we don't have to do that. We can just always set it and then we don't have any issues?

@s1monw Yes, that's correct. Another issue is the validation in CCR that requires an explicit value.

**Comment by jasontedor** [comment]
> Another issue is the validation in CCR that requires an explicit value.

Let’s work to remove this then. I don’t like what you had to do in create index service for example. I’m okay saying if not explicitly set, check the index creating version and assume false if < some cutoff (e.g., 7.0.0 for simplicity) else true.

File: docs/reference/indices/flush.asciidoc [hunk 1]
```diff
@@ -101,7 +101,8 @@ which returns something similar to:
                      "translog_generation" : "2",
                      "max_seq_no" : "-1",
                      "sync_id" : "AVvFY-071siAOuFGEO9P", <1>
-                     "max_unsafe_auto_id_timestamp" : "-1"
+                     "max_unsafe_auto_id_timestamp" : "-1",
+                     "min_retained_seq_no": "0"
```
Comment (bleskes) [inline docs/reference/indices/flush.asciidoc]: why was this added?

Comment (dnhatn) [reply docs/reference/indices/flush.asciidoc]: We [store](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L2318) the min_retained_seq_no of SoftDeletesPolicy in a Lucene index commit.

Comment (bleskes) [reply docs/reference/indices/flush.asciidoc]: But this is not part of this pr?

Comment (dnhatn) [reply docs/reference/indices/flush.asciidoc]: Ah, this change is a part of this PR. We bake this value iff soft-deletes is enabled. Previously, we did not enable soft-deletes but now we do by default; therefore a Lucene commit also by default has it.

**Review by jasontedor — APPROVED** [review]

**Review by s1monw — APPROVED** [review]

**Comment by jasontedor** [comment]
@dnhatn I pushed one more documentation change. It is from documentation that I added in #36333 after you opened this pull request.

**Comment by dnhatn** [comment]
Thanks everyone :).
